### ISSUE TITLE
Box with overflow:hidden cut inner div with absolute position by content edge, not by padding edge

### ISSUE THREAD

**Issue by blacksvk** [servo/servo#16576]
Tested on Windows 8.1 x64

![servo-overflow](https://cloud.githubusercontent.com/assets/10187350/25311540/c25634c6-2803-11e7-9d60-2cd980dff0a5.png)

```
<html>
	<body>
		<style>
			#box{
				background-color: #080b0c;
				height: 100px;
				overflow: hidden;
				padding: 0 15px;
				position: relative;
			}
			#inside {
				position: absolute;
				right: 0px;
				width: 100px;
				height: 100%;
				background: red;
			}
		</style>
		<div id="box">
			<div id="inside">
			</div>
		</div>
	</body>
</html>
```

**Comment by aacunningham** [comment]
What's the best way to recreate these manual testcases? Would it be to create a wpt?

**Comment by jdm** [comment]
@aacunningham It's easiest to create a local html file and point Servo at it. To submit an actual fix, it's recommended to convert it into a [reftest](http://web-platform-tests.org/writing-tests/reftests.html) as part of the automated testsuite.

**Comment by aacunningham** [comment]
After reading through some of the spec on block widths, learning some lingo, and reading a bit of code, I think I'd like to take this on. Do I assign myself with highfive or through you, @jdm?

**Comment by sciguyryan** [comment]
Is it possible that issue #16559 is also related to this since they do seem to have some traits in common?

**Comment by jdm** [comment]
I am doubtful.

**Comment by aacunningham** [comment]
Same, since that issue seems to apply more to the hover action and not an issue with the rendered width of boxes. I'm about to do some debugging today, but I'm going to start my look in the width calculations for absolute nonreplaced blocks and see where that gets me

**Comment by jdm** [comment]
https://github.com/servo/servo/wiki/Getting-started-with-layout may have tips to help you understand the code a bit better.

**Comment by jdm** [comment]
Additionally, we have https://github.com/servo/servo/tree/master/etc/layout_viewer which can help break down what happens in which part of layout for a given testcase.

**Comment by aacunningham** [comment]
Don't think I've read that one yet, thanks!

**Comment by jdm** [comment]
I retract my earlier doubt. @sciguyryan is right; this looks like it's a regression from #16336 as well.

**Comment by jdm** [comment]
cc @mrobinson

**Comment by aacunningham** [comment]
I probably shouldn't have had any doubt since I'm pretty new to this, but also retracted. What's the best way moving forward with this ticket then? Would it make more sense to have a new issue made that's a little more overarching, or is this issue fine? And since mrobinson would be most how the regression was introduced, would it make more sense to have them take over, or is it fine if I keep looking into the issue?

**Comment by jdm** [comment]
Clearly the cause of regression is the same, but the fix for both issues might be different, so I don't think it makes sense to file an overarching one. I don't think there's anything wrong with you continuing to investigate this weekend; if you come up with any conclusions then perhaps @mrobinson could provide useful input on Monday.

**Comment by aacunningham** [comment]
Perfect, thanks a lot!

**Comment by aacunningham** [comment]
Alright, here's what I've discovered:

Using `-Z dump-display-list`, I got this info:

Pre #16336 

```
┌ Display List
│  ├─ Items
│  │  ├─ PushStackingContext(StackingContext at TypedRect(1920px×1043px at (0px,0px)) with overflow TypedRect(1920px×1043px at (0px,0px)): StackingContextId(0)) StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 0, 0, 0) @ TypedRect(1920px×1043px at (0px,0px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 0, 0, 0) @ TypedRect(1904px×100px at (8px,8px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ PushStackingContext(StackingContext at TypedRect(1904px×100px at (8px,8px)) with overflow TypedRect(1904px×100px at (0px,0px)): StackingContextId(139840704336464)) StackingContext: StackingContextId(139840704336464) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0.03137255, 0.043137256, 0.047058824, 1) @ TypedRect(1904px×100px at (0px,0px)) ClippingRegion::Max StackingContext: StackingContextId(139840704336464) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(1, 0, 0, 1) @ TypedRect(100px×100px at (1804px,0px)) ClippingRegion(Rect=TypedRect(1904px×100px at (0px,0px)), Complex=[]) StackingContext: StackingContextId(139840704336608) ScrollRoot: ScrollRootId(0)
│  │  ├─ PopStackingContext(StackingContextId(139840704336464) StackingContext: StackingContextId(139840704336464) ScrollRoot: ScrollRootId(0)
│  │  └─ PopStackingContext(StackingContextId(0) StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
```
Post #16336 

```
┌ Display List
│  ├─ Items
│  │  ├─ PushStackingContext(StackingContext at TypedRect(1920px×1043px at (0px,0px)) with overflow TypedRect(1920px×1043px at (0px,0px)): StackingContextId(0)) StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 0, 0, 0) @ TypedRect(1920px×1043px at (0px,0px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0, 0, 0, 0) @ TypedRect(1904px×100px at (8px,8px)) ClippingRegion::Max StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ PushStackingContext(StackingContext at TypedRect(1904px×100px at (8px,8px)) with overflow TypedRect(1904px×100px at (0px,0px)): StackingContextId(140606955442768)) StackingContext: StackingContextId(140606955442768) ScrollRoot: ScrollRootId(0)
│  │  ├─ PushScrollRoot(ScrollRoot { id: ScrollRootId(140606955442768), parent_id: ScrollRootId(0), clip: ClippingRegion(Rect=TypedRect(1874px×100px at (0px,0px)), Complex=[]), content_rect: TypedRect(1874px×100px at (15px,0px)) } StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(0.03137255, 0.043137256, 0.047058824, 1) @ TypedRect(1904px×100px at (0px,0px)) ClippingRegion::Max StackingContext: StackingContextId(140606955442768) ScrollRoot: ScrollRootId(0)
│  │  ├─ SolidColor rgba(1, 0, 0, 1) @ TypedRect(100px×100px at (1804px,0px)) ClippingRegion::Max StackingContext: StackingContextId(140606955442912) ScrollRoot: ScrollRootId(140606955442768)
│  │  ├─ PopStackingContext(StackingContextId(140606955442768) StackingContext: StackingContextId(140606955442768) ScrollRoot: ScrollRootId(0)
│  │  └─ PopStackingContext(StackingContextId(0) StackingContext: StackingContextId(0) ScrollRoot: ScrollRootId(0)
```
Diffed the two, found the main difference being that the new code adds a `ScrollRoot`. I'm not entirely sure what `ScrollRoot` is, BUT, if it's a representation of the scroll containers referenced in this [CSS spec](https://www.w3.org/TR/css-overflow-3/#valdef-overflow-hidden), then it's appropriate to have `ScrollRoot` added. So correct me if I'm wrong, but I think the issue isn't that `ScrollRoot` is being added, but that there is something wrong in the calculations for `ScrollRoot`; Post #16336, `ScrollRoot`'s `ClippingRegion` is 30px smaller than it should (from the 15px padding).

Going into this, I found that the `ClippingRegion` for the new `ScrollRoot` isn't based on the border box but the content box (which is the border box but with the padding added, not sure if those are technical terms, just variable names). So I modified the code that added the padding and the test case is passing now. I'm not sure if that's the correct solution, but it does work. Trying to run tests, but they run pretty slowly on this machine so it might be easier to have the CI systems handle it.

If none of this seems completely crazy, I'll clean up my code, learn how to add a reftest, and push up a PR.

**Comment by jdm** [comment]
Seems plausible! https://github.com/servo/servo/blob/master/tests/wpt/README.md#writing-new-tests has information about adding a new test.

**Comment by mrobinson** [comment]
```
<html>
  <div id="1" style="overflow:hidden;">
    <div id="2" style="clip: ...">
      <div id="3" style="position:fixed">
      </div>
    </div>
  </div>
</html>
```

```
 
Frame tree (roughly)
 
nsCanvasFrame(html) <
  nsHTMLScrollFrame(1) <
    nsBlockFrame(1) <
      nsBlockFrame(2) <
        nsPlaceholderFrame(3) -> nsBlockFrame(3)
      >
    >
  >
 
>
FixedList <
  nsBlockFrame(3)
>
 ```

Note that we have two different types of clipping here, the clip to make overflow:hidden work, and the clip from the css 'clip' property. These work differently. The overflow clip doesn't apply to the out of flow frame, but the css one does.

To handle this in Gecko, we build and store two clips as we descend into the frame tree: GetClipChainForContainingBlockDescendants (for overflow:hidden) and GetClipChainForContentDescendants (for css clip).
 
We descend into the FixedList before we descend into normal child lists and store the current visible/dirty rect, as well as the current clip for containing block descendants (so without the overflow:hidden clip).
 
When we descend into the normal child lists and eventually reach the placeholder, the current clip state has both of the clips applied. At this point we restore the state we saved earlier (which removes the overflow:hidden clip, but leaves the css one), and then we build items for the position:fixed block frame.

**Comment by mrobinson** [comment]
I have a fix for this at https://github.com/mrobinson/servo/commit/f190445f64921f744946345827bfe9472157ae25. It's just waiting on the WebRender update.
